tests/test_optim.py

```diff
@@ -75,13 +75,13 @@ def test_minimize(tensorlib, precision, optimizer, do_grad, do_stitch):
             # do grad, scipy, 32b
             'do_grad-scipy-pytorch-32b': [0.49993881583213806, 1.0001085996627808],
             'do_grad-scipy-tensorflow-32b': [0.4999384582042694, 1.0001084804534912],
-            'do_grad-scipy-jax-32b': [0.4999389052391052, 1.0001085996627808],
+            'do_grad-scipy-jax-32b': [0.4978247582912445, 1.0006263256072998],
             # do grad, scipy, 64b
             'do_grad-scipy-pytorch-64b': [0.49998837853531425, 0.9999696648069287],
             'do_grad-scipy-tensorflow-64b': [0.4999883785353142, 0.9999696648069278],
             'do_grad-scipy-jax-64b': [0.49998837853531414, 0.9999696648069285],
             # no grad, minuit, 32b - not very consistent for pytorch
-            'no_grad-minuit-numpy-32b': [0.49622172117233276, 1.0007264614105225],
+            'no_grad-minuit-numpy-32b': [0.7465415000915527, 0.8796938061714172],
```

Comment:
this is an incredibly large change

Reply:
Note the comment just below it:

```python
#    nb: macos gives different numerics than CI
# 'no_grad-minuit-pytorch-32b': [0.7465415000915527, 0.8796938061714172]
```

Reply:
Note the comment above it

```
            # no grad, minuit, 32b - not very consistent for pytorch
```

Reply:
Can you clarify why you think this is a problem? We know `32b` is not consistent with `64b`, and now that we're _actually_ using `32b` it isn't surprising to me that this change exists.

Reply:
But now you've managed to get numpy32b in CI to match the pytorch 32b on Mac OS (but also does not match the 32b in CI). So now i'm just generally confused.

Reply:
I'm generally ok with this change, but I need to understand why we've become more inconsistent within just linux alone.

Reply:
> But now you've managed to get numpy32b in CI to match the pytorch 32b on Mac OS (but also does not match the 32b in CI). So now i'm just generally confused.

Ugh. Me too. :(

Reply:
So on `master` (which is partially `32b` and partially `64b`) (with a print statement added)

```
$ python -m pytest -sx tests/test_optim.py -k test_minimize[do_grad-scipy-jax-32b-no_stitch]
=========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.8.6, pytest-6.2.1, py-1.10.0, pluggy-0.13.1
benchmark: 3.2.3 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
Matplotlib: 3.3.3
Freetype: 2.6.1
rootdir: /home/feickert/Code/GitHub/pyhf, configfile: pyproject.toml
plugins: benchmark-3.2.3, console-scripts-1.1.0, anyio-2.1.0, mpl-0.12, mock-3.5.1, cov-2.11.1
collected 161 items / 160 deselected / 1 selected                                                                                                                                                         

tests/test_optim.py [0.4999389052391052, 1.0001085996627808]
.
$ python -m pytest -sx tests/test_optim.py -k test_minimize[no_grad-minuit-numpy-32b-no_stitch]
=========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.8.6, pytest-6.2.1, py-1.10.0, pluggy-0.13.1
benchmark: 3.2.3 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
Matplotlib: 3.3.3
Freetype: 2.6.1
rootdir: /home/feickert/Code/GitHub/pyhf, configfile: pyproject.toml
plugins: benchmark-3.2.3, console-scripts-1.1.0, anyio-2.1.0, mpl-0.12, mock-3.5.1, cov-2.11.1
collected 161 items / 160 deselected / 1 selected                                                                                                                                                         

tests/test_optim.py [0.49622172117233276, 1.0007264614105225]
.
```

and on this branch

```
$ python -m pytest -sx tests/test_optim.py -k test_minimize[do_grad-scipy-jax-32b-no_stitch]
=========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.8.6, pytest-6.2.1, py-1.10.0, pluggy-0.13.1
benchmark: 3.2.3 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
Matplotlib: 3.3.3
Freetype: 2.6.1
rootdir: /home/feickert/Code/GitHub/pyhf, configfile: pyproject.toml
plugins: benchmark-3.2.3, console-scripts-1.1.0, anyio-2.1.0, mpl-0.12, mock-3.5.1, cov-2.11.1
collected 161 items / 160 deselected / 1 selected                                                                                                                                                         

tests/test_optim.py [0.4978247582912445, 1.0006263256072998]
.
$ python -m pytest -sx tests/test_optim.py -k test_minimize[no_grad-minuit-numpy-32b-no_stitch]
=========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.8.6, pytest-6.2.1, py-1.10.0, pluggy-0.13.1
benchmark: 3.2.3 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
Matplotlib: 3.3.3
Freetype: 2.6.1
rootdir: /home/feickert/Code/GitHub/pyhf, configfile: pyproject.toml
plugins: benchmark-3.2.3, console-scripts-1.1.0, anyio-2.1.0, mpl-0.12, mock-3.5.1, cov-2.11.1
collected 161 items / 160 deselected / 1 selected                                                                                                                                                         

tests/test_optim.py [0.7465415000915527, 0.8796938061714172]
.
```

I'm not really sure how much there is to dig in here, other than floating point difference between `64b` finally becoming `32b`.

Thoughts @lukasheinrich @kratsg?

Reply:
We'll have to add more comments in the test and just keep notes to ourselves about these changes...

Reply:
Fair. I think that when we go through PR #1057 some of this will also get further brought up.